File: modules/local/celda/decontx/templates/decontx.py
#!/usr/bin/env python3

import anndata as ad
import anndata2ri
import rpy2
import rpy2.robjects as ro
import platform
celda = ro.packages.importr('celda')

def format_yaml_like(data: dict, indent: int = 0) -> str:
    """Formats a dictionary to a YAML-like string.

    Args:
        data (dict): The dictionary to format.
        indent (int): The current indentation level.

    Returns:
        str: A string formatted as YAML.
    """
    yaml_str = ""
    for key, value in data.items():
        spaces = "  " * indent
        if isinstance(value, dict):
            yaml_str += f"{spaces}{key}:\\n{format_yaml_like(value, indent + 1)}"
        else:
            yaml_str += f"{spaces}{key}: {value}\\n"
    return yaml_str

adata = ad.read_h5ad("${h5ad}")
sc_experiment = anndata2ri.py2rpy(adata)

corrected = celda.decontX(sc_experiment, batch=adata.obs['${batch_col}'].tolist())
counts = celda.decontXcounts(corrected)

adata.layers['ambient'] = anndata2ri.rpy2py(counts).T
adata.X = adata.layers['ambient']
adata.write_h5ad("${prefix}.h5ad")

# Versions

versions = {
    "${task.process}": {
        "python": platform.python_version(),
        "anndata": ad.__version__,
        "anndata2ri": anndata2ri.__version__,
        "rpy2": rpy2.__version__,
        "celda": celda.__version__,
    }
}

with open("versions.yml", "w") as f:
    f.write(format_yaml_like(versions))

File: modules/local/celda/decontx/templates/decontx.R
#!/usr/bin/env Rscript

library(anndataR)
library(celda)

# Read the AnnData object
adata <- read_h5ad("${h5ad}")

# Convert to SingleCellExperiment, mapping only the target input to "counts"
# assays_mapping = FALSE avoids duplicate name conflicts when a "counts" layer already exists
input_layer <- "${input_layer}"
if (input_layer == "X") {
    sce <- adata\$as_SingleCellExperiment(x_mapping = "counts", assays_mapping = FALSE)
} else {
    sce <- adata\$as_SingleCellExperiment(x_mapping = FALSE, assays_mapping = c(counts = input_layer))
}

# Prepare parameters for decontX
params <- list()
params\$assayName <- "counts"

# Handle batch information if available
batch_col <- "${batch_col ?: ''}"

if (batch_col != "") {
    # Plausibility check 1: Does the column exist?
    if (!(batch_col %in% colnames(adata\$obs))) {
        cat("ERROR: Batch column '", batch_col, "' not found in obs columns.\\n", sep="")
        cat("Available columns in obs:", paste(colnames(adata\$obs), collapse=", "), "\\n")
        stop("Batch column '", batch_col, "' does not exist in the AnnData object. Available columns: ", paste(colnames(adata\$obs), collapse=", "))
    }

    batch_values <- adata\$obs[[batch_col]]

    # Plausibility check 2: Does it contain NA/NaN values?
    na_count <- sum(is.na(batch_values))
    if (na_count > 0) {
        unique_vals <- unique(batch_values)
        cat("ERROR: Batch column '", batch_col, "' contains ", na_count, " NA/NaN value(s).\\n", sep="")
        cat("Unique batch values (including NA/NaN):", paste(unique_vals, collapse=", "), "\\n")
        stop("Batch column '", batch_col, "' contains NA/NaN values. All batch values must be non-NA. Unique values: ", paste(unique_vals, collapse=", "))
    }

    # Both checks passed - use the batch column
    params\$batch <- batch_values
}

# Handle background data if available
raw_path <- "${raw}"
if (file.exists(raw_path)) {
    raw <- read_h5ad(raw_path)
    raw_sce <- raw\$as_SingleCellExperiment(x_mapping = "counts", assays_mapping = FALSE)
    params\$background <- raw_sce

    # If a batch column is provided, we also need to add it to the background data
    if (!is.null(params\$batch)) {
        params\$bgBatch <- raw\$obs[[batch_col]]
    }
}

# Run decontX with parameters
corrected <- do.call(decontX, c(list(sce), params))

adata_corrected <- as_AnnData(corrected)

# Convert back to AnnData and update layers
if ("${output_layer}" == "X") {
    adata\$X <- adata_corrected\$layers[["decontXcounts"]]
} else {
    adata\$layers[["${output_layer}"]] <- adata_corrected\$layers[["decontXcounts"]]
}

# Save the output
write_h5ad(adata, "${prefix}.h5ad")

# Write version information
writeLines(
    c(
        '"${task.process}":',
        paste('    r:', paste(version\$major, version\$minor, sep = ".")),
        paste('    anndataR:', as.character(packageVersion('anndataR'))),
        paste('    celda:', as.character(packageVersion('celda')))
    ),
'versions.yml')
